### PR TITLE
Added Share Soft Delete

### PR DESCRIPTION
**Pull request by seanmcc-msft**

### CODE REVIEW

File: sdk/storage/Azure.Storage.Files.Shares/tests/ShareClientTests.cs
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Files.Shares/tests/ShareClientTests.cs:754]: Would be better to wait on/poll (with exponential retry) some API for delete condition instead of hard-coded wait time. Might give more stability in live mode.

File: sdk/storage/Azure.Storage.Files.Shares/tests/ShareClientTests.cs
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Files.Shares/tests/ShareClientTests.cs:779]: nit: can you introduce descriptive constant or variable for that string?

**Review by kasobol-msft — APPROVED** [review]

File: sdk/storage/Azure.Storage.Common/tests/Shared/StorageTestBase.cs
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Common/tests/Shared/StorageTestBase.cs:142]: Copy/Paste doc issue

**Comment by seanmcc-msft** [comment]
@tg-msft, could you look at this PR next?  It's relatively small.

File: sdk/storage/Azure.Storage.Common/tests/Shared/TestConfigurations.cs
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Common/tests/Shared/TestConfigurations.cs:135]: Is Soft Delete an account level setting that applies regardless of whether you're using Blobs/Files?

File: sdk/storage/Azure.Storage.Files.Shares/src/ShareClient.cs
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Files.Shares/src/ShareClient.cs:2072]: Does the operation just fail if you call this with an older service version?  I can't imagine this causing any problems, but want to double check we don't need a version guard here.

File: sdk/storage/Azure.Storage.Files.Shares/tests/ShareClientTests.cs
Comment (tg-msft) [reply sdk/storage/Azure.Storage.Files.Shares/tests/ShareClientTests.cs:754]: Could we just keep calling Delete periodically until we stop getting `ShareBeingDeleted` or whatever the error code is?  If you do keep this as-is, we've got a handy https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/storage/Azure.Storage.Common/tests/Shared/StorageTestBase.cs#L354 that encapsulates the mode check.

**Review by tg-msft — APPROVED** [review]
The only major question I have is about the name (also in ApiView).  We called this `Undelete` for Blobs and we're using `Restore` here.  Any reason we don't want `Undelete` for both?

File: sdk/storage/Azure.Storage.Common/tests/Shared/TestConfigurations.cs
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Common/tests/Shared/TestConfigurations.cs:135]: No, there is Container Soft Delete and Blob Soft Delete for the Blobs service, and Share Soft Delete for Files.  I'd like to just make 1 new test account with all soft delete features enabled.

The reason I want a separate account is that 1. it takes time to enable/disable these features on an account, and 2. if you list with the "Deleted" option, and soft delete has been enabled for weeks, the resulting list could be huge.

File: sdk/storage/Azure.Storage.Files.Shares/src/ShareClient.cs
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Files.Shares/src/ShareClient.cs:2072]: Yes it will fail, with a message from the service saying the version is incorrect.  Peter decided we weren't doing version guarding in the SDK.

**Comment by seanmcc-msft** [comment]
> The only major question I have is about the name (also in ApiView). We called this `Undelete` for Blobs and we're using `Restore` here. Any reason we don't want `Undelete` for both?

If we want to keep the APIs consistent, I'd prefer "Restore" over "Undelete".  For the original naming of the APIs, I just copied the service.

**Comment by seanmcc-msft** [comment]
> Could we just keep calling Delete periodically until we stop getting ShareBeingDeleted or whatever the error code is? If you do keep this as-is, we've got a handy https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/storage/Azure.Storage.Common/tests/Shared/StorageTestBase.cs#L338 that encapsulates the mode check.

I'll attempt to call Delete in a loop until the error code changes and see what happens.

File: sdk/storage/Azure.Storage.Common/tests/Shared/TestConfigurations.cs
Comment (tg-msft) [reply sdk/storage/Azure.Storage.Common/tests/Shared/TestConfigurations.cs:135]: Got it - I'm on board with the separate account.  Just wonder if we should call the tenant name `"TargetBlobsAndFilesSoftDeleteTenant"` or just `"TargetSoftDeleteTenant"` since you need to enable it for both services?  I don't feel very strongly here.

File: sdk/storage/Azure.Storage.Files.Shares/src/ShareClient.cs
Comment (tg-msft) [reply sdk/storage/Azure.Storage.Files.Shares/src/ShareClient.cs:2072]: I thought it was specifically no guards unless it could lead to data corruption or loss - which is why I like double checking for some of these new operations.

**Comment by tg-msft** [comment]
> If we want to keep the APIs consistent, I'd prefer "Restore" over "Undelete". For the original naming of the APIs, I just copied the service.

Should we rename `Undelete` in Blobs to `Restore` (with the old method hidden by `[EditorBrowsable(Never)]` and calling into the new one)?  This is maybe worth having a quick chat with the right folks on your team to make a call here.  It's going to be a little weird to have `ShareClient.Restore` and `BlobContainerClient.Restore` but `BlobClient.Undelete`.  Resolving that question doesn't need to block this PR from getting in.

**Comment by seanmcc-msft** [comment]
> > If we want to keep the APIs consistent, I'd prefer "Restore" over "Undelete". For the original naming of the APIs, I just copied the service.
> 
> Should we rename `Undelete` in Blobs to `Restore` (with the old method hidden by `[EditorBrowsable(Never)]` and calling into the new one)? This is maybe worth having a quick chat with the right folks on your team to make a call here. It's going to be a little weird to have `ShareClient.Restore` and `BlobContainerClient.Restore` but `BlobClient.Undelete`. Resolving that question doesn't need to block this PR from getting in.

Ok, I'll leave the API name as-in for now, and we can circle back around once 73 is closer to the main Arch Board review.

File: sdk/storage/Azure.Storage.Common/tests/Shared/StorageTestBase.cs
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Common/tests/Shared/StorageTestBase.cs:142]: fixed.

File: sdk/storage/Azure.Storage.Common/tests/Shared/TestConfigurations.cs
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Common/tests/Shared/TestConfigurations.cs:135]: I'll add additional documentation around what settings this tenant requres.

File: sdk/storage/Azure.Storage.Files.Shares/src/ShareClient.cs
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Files.Shares/src/ShareClient.cs:2072]: Nothing will happen to your share if you call Restore with a service version that doesn't support it - you will just get a 400 class error.

File: sdk/storage/Azure.Storage.Files.Shares/tests/ShareClientTests.cs
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Files.Shares/tests/ShareClientTests.cs:779]: fixed.

File: sdk/storage/Azure.Storage.Files.Shares/tests/ShareClientTests.cs
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Files.Shares/tests/ShareClientTests.cs:754]: Calling delete multiple tests doesn't work, delete doesn't return an error code if a share is being deleted.  I'll switch to using the Delay method.

Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Files.Shares/tests/ShareClientTests.cs]: I think we can change this before merging to master to calling Restore() in a loop and waiting for no error, for stability in the live tests.

**Pull request merged**
